File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/complexity/counters/JavaCounter.kt
/**
 * KSA-161: Java AST node counter.
 */
package com.codeintel.analyzers.complexity.counters

import com.codeintel.analyzers.complexity.BaseNodeCounter

class JavaCounter : BaseNodeCounter() {
    override val language = "java"
    override val branchNodeTypes = setOf(
        "if_statement", "switch_expression", "ternary_expression",
    )
    override val loopNodeTypes = setOf(
        "for_statement", "enhanced_for_statement",
        "while_statement", "do_statement",
    )
    override val logicalOperators = setOf("&&", "||")
    override val exceptionNodeTypes = setOf("catch_clause")
}

File: backend/src/engine/analyzers/complexity/counters/JavaCounter.ts
/**
 * KSA-161: Java decision point counter.
 */

import type { SyntaxNode } from '../../../parsers/types.js';
import { BaseNodeCounter } from './BaseNodeCounter.js';

export class JavaCounter extends BaseNodeCounter {
  readonly language = 'java';
  readonly branchNodeTypes = [
    'if_statement',
    'switch_expression',
    'ternary_expression',
  ];
  readonly loopNodeTypes = [
    'for_statement',
    'enhanced_for_statement',
    'while_statement',
    'do_statement',
  ];
  readonly logicalOperators = ['&&', '||'];
  readonly exceptionNodeTypes = ['catch_clause'];

  protected isLogicalOp(node: SyntaxNode): boolean {
    if (node.type === 'binary_expression') {
      const op = node.childForFieldName('operator');
      if (op && this.logicalOperators.includes(op.text)) return true;
    }
    return false;
  }
}
